include/nil/actor/zk/commitments/detail/polynomial/proof_of_work.hpp

```diff
@@ -83,6 +87,96 @@ namespace nil {
                         return ((result & mask) == 0);
                     }
                 };
+
+                // Note that the interface here is slightly different from the one above:
+                // amount of bits for grinding instead of the mask.
+                // This was done because the actual mask is applied to the high bits instead of the low bits
+                // which makes manually setting the mask error-prone.
+                template<typename TranscriptHashType, typename FieldType, std::uint8_t GrindingBits=16>
+                class field_proof_of_work {
+                public:
+                    using transcript_hash_type = TranscriptHashType;
+                    using transcript_type = transcript::fiat_shamir_heuristic_sequential<transcript_hash_type>;
+                    using value_type = typename FieldType::value_type;
+                    using integral_type = typename FieldType::integral_type;
+
+                    constexpr static const integral_type mask =
+                        (GrindingBits > 0 ?
+                            ((integral_type(2) << GrindingBits - 1) - 1) << (FieldType::modulus_bits - GrindingBits)
+                            : 0);
+
+                    static inline boost::property_tree::ptree get_params() {
+                        boost::property_tree::ptree params;
+                        params.put("mask", mask);
+                        return params;
+                    }
+
+                    static inline value_type generate(transcript_type &transcript, 
```

Comment:
Let's make this "const transcript_type &transcript".

Reply:
`transcript` could not be made `const`, because the grinded value should be passed into it and challenge queried.
https://github.com/NilFoundation/actor-zk/pull/35/files#diff-01026ab5ff0e47443c8145d5bb17b33459c18e46ad308ba8a5528e202965de7eR144-R146

Reply:
I did not notice it. Can you please divide this function into 2 parts, one that find the POW, and another one that passes it into the transcript? 

This way the upper function can have a const transcript, also this part

auto it = results.begin();
  while (it != results.end() ) {
      std::optional<value_type> result = it->get();
      if (result) {
          pow_value = *result;
          break;
      }
      ++it;
  }
                            
  can be written as
  
  for (const auto& v: results) {
   std::optional<value_type> result = v.get();
  if (result) {
  return *result;
  }
  }